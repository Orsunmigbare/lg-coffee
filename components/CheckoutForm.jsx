import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { CardElement, useElements, useStripe } from '@stripe/react-stripe-js';
import axios from 'axios';
import Form from 'react-bootstrap/Form';
import BillingDetailsFields from './BillingDetailsFields';
import SubmitButton from './SubmitButton';
import CheckoutError from './CheckoutError';

const handleUSDChange = (number) => {
  const string = JSON.stringify(number);
  const { length } = string;
  return JSON.parse(
    `${string.substring(0, length - 2)}.${string.substring(length - 2)}`,
  );
};

const CheckoutForm = ({ price, onSuccessfulCheckout }) => {
  const [isProcessing, setProcessingTo] = useState(false);
  // eslint-disable-next-line no-unused-vars
  const [checkoutError, setCheckoutError] = useState();
  // eslint-disable-next-line no-unused-vars
  const [total, setTotal] = useState();
  const [USD, setUSD] = useState();

  useEffect(() => {
    setTotal(price);
    if (total > 0) {
      setUSD(handleUSDChange(total));
    }
  }, [price, total]);

  console.log(USD);

  const stripe = useStripe();
  const elements = useElements();

  const handleFormSubmit = async (ev) => {
    ev.preventDefault();

    const billingDetails = {
      name: ev.target.name.value,
      email: ev.target.email.value,
      address: {
        city: ev.target.city.value,
        line1: ev.target.address.value,
        state: ev.target.state.value,
        postal_code: ev.target.zip.value,
      },
    };

    setProcessingTo(true);

    // create payment intent
    const { data: clientSecret } = await axios.post('/api/payment_intents', {
      amount: price,
    });

    const cardElement = elements.getElement(CardElement);

    const paymentMethodReq = await stripe.createPaymentMethod({
      type: 'card',
      card: cardElement,
      billing_details: billingDetails,
    });

    // eslint-disable-next-line no-unused-vars
    const confirmCardPayment = await stripe.confirmCardPayment(clientSecret, {
      payment_method: paymentMethodReq.paymentMethod.id,
    });

    onSuccessfulCheckout();
  };

  const cardElementOptions = {
    style: {
      base: {
        iconColor: '#c4f0ff',
        color: '#fff',
        fontWeight: 500,
        fontFamily: 'Roboto, Open Sans, Segoe UI, sans-serif',
        fontSize: '16px',
        fontSmoothing: 'antialiased',
        ':-webkit-autofill': {
          color: '#fce883',
        },
        '::placeholder': {
          color: '#87BBFD',
        },
      },
      invalid: {
        iconColor: '#FFC7EE',
        color: '#FFC7EE',
      },
    },
  };
  return (
    <Form onSubmit={handleFormSubmit}>
      <BillingDetailsFields />
      <div
        style={{
          backgroundColor: 'none !important',
          border: '0',
        }}
      >
        <CardElement options={cardElementOptions} />
      </div>

      {checkoutError && <CheckoutError>{checkoutError}</CheckoutError>}

      <SubmitButton
        disabled={isProcessing}
        title={isProcessing ? 'Processing...' : `Pay $${USD || price} USD`}
        className="mt-4 btn-outline-light"
      />
    </Form>
  );
};

CheckoutForm.propTypes = {
  price: PropTypes.number.isRequired,
  onSuccessfulCheckout: PropTypes.func.isRequired,
  // eslint-disable-next-line react/require-default-props
  cart: PropTypes.shape({}),
};

export default CheckoutForm;
