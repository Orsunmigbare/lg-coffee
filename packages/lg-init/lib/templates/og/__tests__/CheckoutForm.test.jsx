import React from 'react';
import ReactDOM from 'react-dom';
import Router from 'next/router';
import renderer from 'react-test-renderer';
import StripeWrapper from '../components/StripeWrapper';
import CheckoutForm from '../components/CheckoutForm';

it('CheckoutForm component renders smoke test without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <StripeWrapper>
      <CheckoutForm price={123} onSuccessfulCheckout={() => Router.push('/checkout/order-confirmation')} />
    </StripeWrapper>,
    div,
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('CheckoutForm component renders unchanged', () => {
  const tree = renderer.create(
    <StripeWrapper>
      <CheckoutForm price={123} onSuccessfulCheckout={() => Router.push('/checkout/order-confirmation')} />
    </StripeWrapper>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
