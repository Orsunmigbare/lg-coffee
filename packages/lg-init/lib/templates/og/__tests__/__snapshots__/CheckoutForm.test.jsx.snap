// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm component renders unchanged 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="d-flex flex-column align-items-center"
  >
    <div
      style={
        Object {
          "maxWidth": "80%",
        }
      }
    >
      <div
        className="form-row"
      >
        <h2
          className="mt-5"
        >
          Billing Information
        </h2>
        <input
          className="mt-3 form-control"
          label="name"
          name="name"
          placeholder="Jane Doe"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
        <input
          className="mt-3 form-control"
          label="email"
          name="email"
          placeholder="jane.doe@example.com"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
        <input
          className="mt-3 form-control"
          label="address"
          name="address"
          placeholder="185 Berry St. Suite 550"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
        <input
          className="mt-3 form-control"
          label="city"
          name="city"
          placeholder="San Francisco"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
        <input
          className="mt-3 form-control"
          label="state"
          name="state"
          placeholder="CA"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
        <input
          className="my-3 form-control"
          label="zip"
          name="zip"
          placeholder="94103"
          required={true}
          style={
            Object {
              "border": "0",
            }
          }
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    className="px-5 py-2"
    style={
      Object {
        "backgroundColor": "none !important",
        "border": "0",
      }
    }
  >
    <div />
  </div>
  <button
    className="my-4 mx-5 btn-outline-dark btn"
    type="submit"
  >
    Pay $undefined USD
  </button>
</form>
`;
